File: C++/代码随想录/Binary_Tree/Find the attributes of Binary tree/110. Balanced Binary Tree.cpp
// 110. Balanced Binary Tree
// https://leetcode.com/problems/balanced-binary-tree/
// clang++ 110.\ Balanced\ Binary\ Tree.cpp -std=c++17
#include <cmath>
#include "../../printUtils.hpp"

template <typename T>
class Solution {
public:
    bool isBalanced(TreeNode<T>* root) {
        if (root == nullptr) {
            return true;
        }

        int leftHeight = getHeight(root->left);
        int rightHeight = getHeight(root->right);

        if (std::abs(leftHeight - rightHeight) > 1) {
            return false;
        }

        return isBalanced(root->left) && isBalanced(root->right);
    }

private:
    int getHeight(TreeNode<T>* node) {
        if (node == nullptr) {
            return 0;
        }

        int leftHeight = getHeight(node->left);
        int rightHeight = getHeight(node->right);

        return 1 + std::max(leftHeight, rightHeight);
    }
};

int main() {
    TreeNode<int>* root = buildTree({3, 9, 20, -1, -1, 15, 7}, -1);
    std::cout << "Is the binary tree balanced? " << std::boolalpha << Solution<int>().isBalanced(root) << std::endl;
    deleteTree(root);
    
    return 0;
}

File: C++/代码随想录/Binary_Tree/Find the attributes of Binary tree/110. Balanced Binary Tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def check(node):
            if not node:
                return 0, True
            left_height, left_balanced = check(node.left)
            right_height, right_balanced = check(node.right)
            return max(left_height, right_height) + 1, left_balanced and right_balanced and abs(left_height - right_height) <= 1
        
        return check(root)[1]

# Test case
root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
sol = Solution()
print(sol.isBalanced(root))  # True
